libr/bin/format/elf/elf.c

```diff
@@ -1277,454 +1296,499 @@ ut64 Elf_(r_bin_elf_get_section_addr_end)(ELFOBJ *bin, const char *section_name)
-	nrel = bin->rel_cache->count;
+	return addr;
+}
 
-	switch (bin->ehdr.e_machine) {
-	case EM_PPC:
-	case EM_PPC64:
-		return get_import_addr_ppc (bin, rel, plt_section, nrel);
-	case EM_SPARC:
-	case EM_SPARCV9:
-	case EM_SPARC32PLUS:
-		return get_import_addr_sparc (bin, rel, plt_section);
-	case EM_RISCV:
-		return get_import_addr_riscv (bin, rel, plt_section);
+static bool is_thumb_symbol(ut64 plt_addr) {
+	return plt_addr & 1;
+}
+
+static ut64 get_import_addr_arm(ELFOBJ *bin, RBinElfReloc *rel) {
+	ut64 got_addr = bin->dyn_info.dt_pltgot;
+	if (!got_addr) {
+		return UT64_MAX;
+	}
+
+	ut64 plt_addr = get_got_entrie (bin, rel);
+	if (plt_addr == UT64_MAX) {
+		return UT64_MAX;
+	}
+
+	ut64 pos = (rel->rva - got_addr - 0x3 * WORDSIZE) / WORDSIZE;
+
+	switch (rel->type) {
+	case R_ARM_JUMP_SLOT: {
+		plt_addr += pos * 12 + 20;
+		if (is_thumb_symbol (plt_addr)) {
+			plt_addr--;
+		}
+		return plt_addr;
+	}
+	case R_AARCH64_JUMP_SLOT:
+		return plt_addr + pos * 16 + 32;
+	default:
+		bprintf ("Unsupported relocation type for imports %d\n", rel->type);
+		return UT64_MAX;
+	}
+	return UT64_MAX;
+}
+
+static ut64 get_import_addr_mips(ELFOBJ *bin, RBinElfReloc *rel) {
+	ut64 addr_jmprel = bin->dyn_info.dt_jmprel;
+	ut64 addr_got = bin->dyn_info.dt_mips_pltgot;
+
+	if (!addr_jmprel || !addr_got) {
+		return UT64_MAX;
+	}
+
+	ut64 pos = (rel->rva - addr_got - 0x2 * WORDSIZE) / WORDSIZE;
+
+	ut8 buf[1024];
+	const ut8 *base;
+	ut64 plt_addr = addr_jmprel + bin->dyn_info.dt_pltrelsz;
+	ut64 p_plt_addr = Elf_(r_bin_elf_v2p_new) (bin, plt_addr);
+	int res = r_buf_read_at (bin->b, p_plt_addr, buf, sizeof (buf));
+	if (res != sizeof (buf)) {
+		return UT64_MAX;
+	}
+
+	base = r_mem_mem_aligned (buf, sizeof (buf), (const ut8 *)"\x3c\x0f\x00", 3, 4);
+	if (base) {
+		plt_addr += (int)(size_t) (base - buf);
+	} else {
+		plt_addr += MIPS_PLT_OFFSET + 8; // HARDCODED HACK
+	}
+	plt_addr += pos * 16;
+
+	return plt_addr;
+}
+
+static ut64 get_import_addr_ppc(ELFOBJ *bin, RBinElfReloc *rel) {
+	ut64 plt_addr = bin->dyn_info.dt_pltgot;
+	ut64 p_plt_addr = Elf_(r_bin_elf_v2p_new) (bin, plt_addr);
+
+	ut8 buf[4] = { 0 };
+	if (p_plt_addr == UT64_MAX) {
+		return UT64_MAX;
+	}
+
+	ut64 nrel = get_num_relocs_dynamic_plt (bin);
+	ut64 pos = (rel->rva - plt_addr) / 4;
+
+	int res = r_buf_read_at (bin->b, p_plt_addr, buf, sizeof (buf));
+	if (res < 4) {
+		return UT64_MAX;
+	}
+
+	if (bin->endian) {
+		ut64 base = r_read_be32 (buf);
+		base -= (nrel * 16);
+		base += (pos * 16);
+		return base;
+	}
+
+	ut64 base = r_read_le32 (buf);
+	base -= (nrel * 12) + 20;
+	base += (pos * 8);
+	return base;
+}
+
+static ut64 get_import_addr_riscv(ELFOBJ *bin, RBinElfReloc *rel) {
+	ut64 got_addr = bin->dyn_info.dt_pltgot;
+	if (!got_addr) {
+		return UT64_MAX;
+	}
+
+	ut64 plt_addr = get_got_entrie (bin, rel);
+	if (plt_addr == UT64_MAX) {
+		return UT64_MAX;
+	}
+
+	ut64 pos = (rel->rva - got_addr - 0x2 * WORDSIZE) / WORDSIZE;
+	return plt_addr + MIPS_PLT_OFFSET + pos * 0x10;
```

Comment:
why are you using `MIPS_PLT_OFFSET` considering the function is `get_import_addr_riscv`? I think it was previously using `RISCV_PLT_OFFSET`.